### PR TITLE
Multi-cell selection, copy, cut, and paste

### PR DESCRIPTION
**Pull request by legnes**
This PR adds multi-cell selection to the builder! It also extends delete and paste to handle selected regions, and adds new cut and copy functionality for both single cells and selected regions

You can control multi-cell selection in a few different ways:
- click and drag
- shift-click
- shift + arrow keys

Closes https://github.com/crosshare-org/crosshare/issues/476

Screenshot 2024-04-04 at 7 13 09 PM

Screenshot 2024-04-04 at 7 28 43 PM

### CODE REVIEW

**Comment by mdirolf** [comment]
Haven't had a real chance to review this yet but from a quick glance it looks really good. Thank you for tackling it! Will try to review tomorrow or this weekend at latest.

File: app/lib/types.ts [hunk 1]
```diff
@@ -119,6 +122,14 @@ export interface PosAndDir extends Position {
   dir: Direction;
 }
 
+export function asPosition({ row, col }: PosAndDir): Position {
```
Comment (mdirolf) [inline app/lib/types.ts:125]: I think maybe this is dead code - don't see it being used anywhere. Don't really see the point to it anyway as anything in TS that takes a Position should take a PosAndDir just fine.

File: app/lib/domUtils.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export function isTextInput(target: EventTarget | null): boolean {
```
Comment (mdirolf) [inline app/lib/domUtils.ts:1]: Love this refactoring!

File: app/lib/types.ts (hunk 1 above)
Comment (mdirolf) [reply app/lib/types.ts:125]: Heh nvm I see where it's used now. Is it really necessary, wouldn't just using the PosAndDir directly work too?

**Review by mdirolf — COMMENTED** [review]
Overall this code looks great.

My one question is that I found the cut/paste behavior didn't work as expected. When I selected a rectangle and cut and pasted somewhere else I expected the paste to kind of maintain that rectangular shape (and any black squares), rather than just filling cells in order. In my mind I'd be able to select an entire corner of the grid and then shift that corner over to a different area. I guess making that change would require a shift to how cut(/copy) works in that we'd need a row deliminator and a column deliminator. What do you think - I'm open to arguments that the way you've got it implemented is the correct way.

As a follow-up (if we decide to do that) we might need to tweak the cell selection style so that it's visible when a black square is selected. We could use something like the style that gets used when a black square is the active cell.

**Comment by legnes** [comment]
> What do you think - I'm open to arguments that the way you've got it implemented is the correct way.

No I think you're totally right! I went back and forth on this and ended up erring on the side of less scope/complexity since I haven't contributed before. But I agree the structured paste with separate row/col delimiters is the more correct behavior. I also really like the idea of being able to cut/copy block squares, I hadn't thought of that!

I'll take a crack at it soon and update this PR if that's ok?

**Comment by mdirolf** [comment]
That sounds great! I'm looking forward to seeing it in action.

File: app/lib/types.ts (hunk 1 above)
Comment (legnes) [reply app/lib/types.ts]: Removed!

**Comment by legnes** [comment]
Pushed some updates!
- Selection paste maintains its shape
- Selection cut/copy/paste can include block squares
- Selected block squares use stripy style
- Shift + arrow keys get their own key enum entries (seemed like this is how other shift combos work)

I debated including bars in copy/paste, but you can't backspace/delete them so I figured they're maybe just a little different? Not sure, I'm open to thoughts/suggestions. 

Let me know what you think!

New style:
Screenshot

**Comment by mdirolf** [comment]
This looks amazing - been playing around with it locally and I think you've pretty much nailed it IMO. I agree with punting on bar handling, at least for now. Not so many folks even use bars to begin with and I think it's safe to wait and see if requests start coming in to add that behavior.

I *think* the way you've implemented it is best but just wanted to quickly get your thoughts - do you think it makes sense to update symmetric blocks when cut/paste operations change a block status? I see that's the way it works currently but I just can't decide if that's best or if it's best for cut/paste to operate w/o honoring any symmetry status (so that they don't unintentionally change remote parts of the grid).

One quick nit - do you mind changing the name `--highlight` to `--selected-cell` or something like that? There are so many color definitions and I already frequently get a bit confused when I look through that list as the names I've chosen are too generic. Otherwise I'm going to give a quick once-over but I'd like to try to get this merged and deployed ASAP - it is awesome!

**Comment by legnes** [comment]
> I think the way you've implemented it is best but just wanted to quickly get your thoughts - do you think it makes sense to update symmetric blocks when cut/paste operations change a block status? I see that's the way it works currently but I just can't decide if that's best or if it's best for cut/paste to operate w/o honoring any symmetry status (so that they don't unintentionally change remote parts of the grid).

I'm glad you noticed this too! It really threw me for a loop when I pasted something in the middle row, and some of the pasted block cells were getting destroyed by pasted letters + symmetry.

I think this might be one of those things that we need more testing/input on?

My instinct is to strictly preserve symmetry, since it's sort of a promise the software is making, and since constructors can turn symmetry off if they want. But I wonder if there's some policy we can adopt that preserves symmetry while limiting unintentional changes, like "don't paste in a cell if it would change the block/non-block status of the symmetrical cell"?

I guess the friendliest thing might be to add some sort of indicator (shading/bg color) that points out symmetry violations, or that pops up when you've pasted a block over a letter via symmetry, though that expands the scope a fair bit.

So I guess if the options are:
A) Paste ignores symmetry
B) Paste respects symmetry
C) Paste respects symmetry by compromising on paste behavior
D) Add a visual clue about symmetry/lack thereof during paste

I'd probably vote B or C as an initial guess. I'm not totally sure though, what do you think? Anything I'm missing?

**Comment by mdirolf** [comment]
> I'm glad you noticed this too! It really threw me for a loop when I pasted something in the middle row, and some of the pasted block cells were getting destroyed by pasted letters + symmetry.

Heh I didn't even think of that case! I think that pushes me towards option A that you've outlined - my gut is that when you paste the number one priority should be to actually transfer all of the cells from the clipboard, even if it violates symmetry.

> So I guess if the options are: A) Paste ignores symmetry B) Paste respects symmetry C) Paste respects symmetry by compromising on paste behavior D) Add a visual clue about symmetry/lack thereof during paste
> 
> I'd probably vote B or C as an initial guess. I'm not totally sure though, what do you think? Anything I'm missing?

I think I'm on team A or B. There is a slight precedence for symmetry being allowed to be violated in some cases - notably if you turn symmetry off and edit a grid and then turn it back on (or import an asymmetric grid while symmetry is turned on) the constructor allows you to continue editing and will make your new changes symmetric w/o worrying about what's already in place. Maybe that's sort of similar to the case of pasting? Or maybe we should just agree that B is correct, especially since that's what's implemented already 😂

**Comment by mdirolf** [comment]
Thank you for the style name changes! Other than the symmetry discussion I think this is ready to merge. Likely won't be another deploy going out until Wednesday, though, so there isn't a huge rush to figure out the symmetry thing.

**Comment by legnes** [comment]
I'm happy to trust your gut and go with A! Especially if there's precedent for breaking symmetry, I think it makes sense to go with the option where you're less likely to accidentally delete/overwrite something.

**Comment by legnes** [comment]
Pushed up the change for this! I made both cut and paste ignore symmetry, though I'm realizing we only really talked about paste. Let me know if you think cut should work differently! Sorry to get so far in the weeds haha

**Comment by mdirolf** [comment]
Thank you for going into the weeds! IMO that's how good software is written. Everything looks great to me. I'm very impressed w/ this, especially as a first contribution! Will try to get a deploy out ASAP (likely Wednesday) and let you know when it's live.

**Comment by mdirolf** [comment]
Deployed! Thank you again for such an excellent contribution. I can't find an email to use but FYI I invited you to the crosshare-org as a member - I don't believe this gives any real privileges but serves as sort of a "badge of honor". Also if you're on the crosshare discord (link on bottom of homepage) msg me there and I'll give you the "developer" role.

**Comment by adamthedog** [comment]
Thanks so much for the contribution! I haven't played with it just yet, but the impl looks excellent.
Cheers!

**Comment by adamthedog** [comment]
One note is we may want a SolverPreferencesList option to avoid overwriting cells containing text with empty cells from the copied region.

**Comment by mdirolf** [comment]
> One note is we may want a SolverPreferencesList option to avoid overwriting cells containing text with empty cells from the copied region.

That’s an interesting thought! My inclination is to wait to hear from some constructors how they are using this new feature first. But we also could open an issue now to track interest if you want

**Comment by legnes** [comment]
:D Thanks! Very exciting!

**Pull request merged**